Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+
+Here's a common pitfall that is seen. Let's continue the menu example, in this example the menu is going to be rendered
+using the current page's root node:
+
+```csharp
+<ul>
+	<li><a href="@Model.Root().Url()">@Model.Root().Name</a></li>
+	@foreach (var node in Model.Root().Children)
+	{
+		<li><a href="@node.Url()">@node.Name</a></li>
+	}
+</ul>
+```
+
+The syntax `@Model.Root()` is shorthand for doing this:
+`Model.AncestorOrSelf(1)` which means it is going to traverse up the tree until it reaches an ancestor node
+with a level of one. As mentioned above, traversing costs resources and in this example there is 3x traversals being done for the same value. Instead this can be rewritten as:
+
+```csharp
+@{
+	var root = Model.Root();
+}
+<ul>
+	<li><a href="@root.Url()">@root.Name</a></li>
+	@foreach (var node in root.Children)
+	{
+		<li><a href="@node.Url()">@node.Name</a></li>
+	}
+</ul>
+```
+
+## Using the Services layer in your views
+
+The Services layer of Umbraco is for manipulating the business logic of Umbraco directly to/from the database.
+None of these methods should be used within your views and can have a very large impact on performance and stability of
+your application.
+
+Your views should rely only on the read-only data services such as `UmbracoHelper`, `ITagQuery` and `IMemberManager` and the properties/methods that they expose. This ensures
+that the data being queried is fast (comes from cache) and that you aren't inadvertently making database changes.
+
+__For example__, when retrieving a content item in your views:
+
+```csharp
+@using Umbraco.Cms.Core.Services
+@inject IContentService _contentService
+
+@{
+	// Services access in your views :(
+	var dontDoThis = _contentService.GetById(1234);
+	
+	// Content cache access in your views
+	var doThis = Umbraco.Content(1234);
+}
+```
+
+If you are using services in your views, you should figure out why this is being done and, in most cases, remove this logic.
+
+## Using Umbraco content items for volatile data
+
+This is one of the worst Umbraco anti-patterns and could very well cause your site to perform ultra poorly.
+
+Umbraco's content should not be used for volatile data, Umbraco's APIs and the way Umbraco's data is persisted
+was never designed for this. If you need to store/write/track data that changes a lot you should use a
+custom database table or another service but not Umbraco content nodes.
+
+Some examples of what not to do are:
+
+* Hit counters to track the number of times your page has been viewed - use something like Google Analytics for this or a custom database table
+* Creating new nodes for form submissions - this should be stored in a custom database table
+* Importing lots of data into Umbraco content nodes that could be stored in a custom database table (i.e. it's not going to be edited).
+In some cases this might be ok but many times we've seen bulk imports occur on an hourly/daily schedule which is generally unnecessary.
```

Comment:
```suggestion
## Using Umbraco content items for volatile data

This is one of the worst Umbraco anti-patterns and could very well cause your site to perform ultra poorly.

Umbraco's content should not be used for volatile data, Umbraco's APIs and the way Umbraco's data is persisted
was never designed for this. If you need to store/write/track data that changes a lot you should use a
custom database table or another service but not Umbraco content nodes.

Some examples of what not to do are:

* Hit counters to track the number of times your page has been viewed - use something like Google Analytics for this or a custom database table
* Creating new nodes for form submissions - this should be stored in a custom database table
* Importing lots of data into Umbraco content nodes that could be stored in a custom database table (i.e. it's not going to be edited).
In some cases, this might be ok but many times we've seen bulk imports occur on an hourly/daily schedule which is generally unnecessary.
```